### PR TITLE
added method to convert circle edges to spline edges

### PR DESCRIPTION
**Pull request by shimwell**
## Proposed changes

as described in issue #737 circle edges can cause problems when importing the stp files into trelis.

This PR provides the user with a method of replacing the circles with spines of specified tolerance.

TODO

- [x] adding tests
- [x] adding docstrings
- [x] perhaps move from shape.py to mixed shapes or utils

## Types of changes

What types of changes does your code introduce to the Paramak?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [x] Documentation Update (if none of the other choices apply)
- [x] New tests

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Pep8 applied
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by shimwell** [comment]
Example usage case. This will convert circles to splines and add extra points so the splines accurately reflect the circle shape. Default tolerance is 0.1 but can be user specified

```
rotated_mixed = paramak.RotateMixedShape(
    rotation_angle=180,
    points=[
        (100, 0, "straight"),
        (200, 0, "circle"),
        (250, 50, "circle"),
        (200, 100, "straight"),
        (150, 100, "straight"),
        (140, 75, "circle"),
        (110, 45, "circle"),
    ]
)
rotated_mixed.convert_all_circle_points_to_splines(tolerance=0.01)
```

**Comment by shimwell** [comment]
tests, doc strings and reorganisation is finished. All the checks and tests are passing.

I think this is ready to be merged. If there are no objects i can merge this in on Tuesday at midday

**Review by RemDelaporteMathurin — COMMENTED** [review]
@Shimwell This looks tidy. Is there a way to test that transforming circles to edges doesn't modify the shape too much ? For instance by creating a circle shape, converting it to splines, and check that the difference in volumes is low ?

**Comment by RemDelaporteMathurin** [comment]
> @Shimwell This looks tidy. Is there a way to test that transforming circles to edges doesn't modify the shape too much ? For instance by creating a circle shape, converting it to splines, and check that the difference in volumes is low ?
> 
> Great idea @RemDelaporteMathurin I was silly and testing for the code coverage more than actually designing good tests. I shall add the suggested test, thanks for the idea

@Shimwell seems like you've been editing my comment rather than replying to it therefore I hadn't seen it ;-)

**Comment by shimwell** [comment]
Great idea @RemDelaporteMathurin I was silly and testing for the code coverage more than actually designing good tests. I shall add the suggested test, thanks for the idea

Screen shot shows the circle edge from the test shape overlaid with the spine replacement edge. This is with the default tolerance (0.1) and there is very little difference (less than 0.000001 rel difference in volume)
Screenshot from 2021-02-27 13-25-44


@eepeterson @jbae11 looks like circles can easily be replaced with splines and the volumes kept to approximately the same value. Hoping this fixes #737

But perhaps more interestingly we could use similar a edge replacement method to find splines and replaced them with a series of straight lines and make spline free geometry for easier conversion to CSG. Perhaps something for the future

File: tests/test_utils.py [hunk 1]
```diff
@@ -12,6 +12,33 @@
...
+        new_points = paramak.utils.convert_circle_to_spline(
+            p0=(200, 0),
+            p1=(250, 50),
+            p2=(200, 100),
+            tolerance=0.2
+        )
+
+        connections = [connection[2] for connection in new_points]
+
+        assert len(set(connections)) == 1
```
Comment (shimwell) [inline tests/test_utils.py]:
```suggestion
        assert new_points[0][0] == 200
        assert new_points[0][1] == 0
        assert new_points[-1][0] == 200
        assert new_points[-1][1] == 100
        assert len(set(connections)) == 1
```

**Review by RemDelaporteMathurin — COMMENTED** [review]
I wonder if the tests should live in test_parametric_shapes since they seem to apply to RotateMixedShape and ExtrudeMixedShape and not only to the super class Shape ?

File: paramak/utils.py [hunk 2]
```diff
@@ -654,6 +654,45 @@ def export_wire_to_html(
...
+
+    new_edge = paramak.utils._transform_curve(edge, tolerance=tolerance)
+
+    points = paramak.utils.extract_points_from_edges(
+        edges=new_edge,
+        view_plane='XZ'
+    )
+    points_with_connections = []
+    for point in points:
+        points_with_connections.append((point[0], point[1]))
```
Comment (RemDelaporteMathurin) [inline paramak/utils.py]: what's the difference between `points` and `points_with_connections` ?

**Review by RemDelaporteMathurin — APPROVED** [review]

File: paramak/utils.py (hunk 2 above)
Comment (shimwell) [reply paramak/utils.py]: ah  yes, sorry I think they were different early on in this PR but now they are the same. I can remove this.

**Comment by shimwell** [comment]
> I wonder if the tests should live in test_parametric_shapes since they seem to apply to RotateMixedShape and ExtrudeMixedShape and not only to the super class Shape ?

I've moved the tests over now and I agree it is better than before

**Pull request merged**